common/filesystem.h

```diff
@@ -0,0 +1,1374 @@
+#include "common/raw_string_ostream.h"
+#include "common/template_string.h"
+#include "llvm/ADT/ScopeExit.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/FormatVariadic.h"
+
+// Provides a filesystem library for use in the Carbon project.
+//
+// This library provides an API designed to support modern Unix / Linux / POSIX
+// style filesystem operations efficiently and securely, while also carefully
+// staying to a set of abstractions and operations that can be reasonably
+// implemented even on Windows platforms.
+//
+// TODO: Currently, there is not a Windows implementation, but this is actively
+// desired when we have testing infrastructure in place for Windows development.
+// Lacking that testing infrastructure and a full Windows port, the operations
+// here are manually compared with LLVM's filesystem library to ensure a
+// reasonable Windows implementation is possible.
+//
+// The library uses C++'s `std::filesystem::path` as its abstraction for paths,
+// and provides two core APIs: open directories and files.
+//
+// Open directories provide relative- and absolute-path based operations to open
+// other directories or files. This allows secure creation of directories even
+// in the face of adversarial operations for example in a shared `/tmp`
+// directory. There is a `constexpr` current working directory available as
+// `Cwd()` that models normal filesystem operations with paths.
+//
+// Open files provide read, write, and other operations on the file. There are
+// separate types for read-only, write-only, and read-write files to model the
+// different APIs available.
+//
+// The APIs for both directories and files are primarily on `*Ref` types that
+// model a non-owning reference to the directory or file. These types are the
+// preferred types to use on an API boundary. Owning versions are provided that
+// ensure the file or directory is closed on destruction. Files support explicit
+// closing in order to observe any close-specific errors.
+//
+// Where APIs require flag parameters of some form, this library provides
+// enumerations that model those flags. The enumeration values are in turn
+// chosen to simplify passing these to specific native APIs. This means the
+// enumeration *values* should not be expected to be portable across platforms.
+// Customizing the values is part of the larger TODO to port the implementation
+// to Windows.
+namespace Carbon::Filesystem {
+
+// Enumerates the different open access modes available. These are largely used
+// to parameterize types in order to constrain which API subset is available.
+enum class OpenAccess {
+  ReadOnly = O_RDONLY,
+  WriteOnly = O_WRONLY,
+  ReadWrite = O_RDWR,
+};
+
+// The different creation styles available when opening a file or directory.
+//
+// Because these are by far the most common parameters and they have unambiguous
+// names, the enumerators are also available directly within the namespace.
+enum class CreationFlags {
```

Comment:
```suggestion
enum class CreateFlags {
```
Grammatically, we have "OpenFlags" so "CreateFlags" would better match I think. Or "CreationFlags" and "OpeningFlags", but the shorter seems better.

I probably would also not use `Flags` for a name that isn't bitflags that can be combined together (the values inside are flags but those are implementation details). Maybe `CreateOptions`, but not critical.. up to you.

Reply:
Yeah, I like that distinction. I think adopted and adjusted accordingly.